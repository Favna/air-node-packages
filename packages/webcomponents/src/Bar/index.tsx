import { Box } from '@material-ui/core';
import classnames from 'classnames';
import React, { memo, PropsWithChildren } from 'react';
import css from './Bar.scss';

interface BarProps {
  /** Additional CSS classes to apply to the Bar component */
  classes?: string | string[];
}

/**
 * Constructs a horizontal Bar section using pre-defined Rijkswaterstaat styling
 * @example
 * ```jsx
 * <Bar>
 *   <div> Some Children </div>
 * </Bar>
 * ```
 */
export const Bar = memo(({ classes, children, ...props }: PropsWithChildren<BarProps>) => (
  <Box {...props} component='div' className={classnames(css.bar, classes)}>
    {children}
  </Box>
));
