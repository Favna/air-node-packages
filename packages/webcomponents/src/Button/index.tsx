import { createStyles, makeStyles, Theme } from '@material-ui/core';
import MUIButton, { ButtonProps as MUIButtonProps } from '@material-ui/core/Button';
import clsx from 'clsx';
import React, { forwardRef, memo, ReactNode } from 'react';
import css from './Button.scss';

export interface ButtonProps {
  /** The variant of the button */
  variant: 'text' | 'outlined' | 'contained' | undefined;
  /** The color type of the button */
  color: 'primary' | 'secondary' | 'default' | 'inherit' | undefined;
  /** The label to be displayed inside the button */
  label: ReactNode;
  /** Whether this button should be disabled */
  disabled?: boolean;
  /** Data-qa tag to apply to the search bar and input element */
  'data-qa'?: string;
  /** Custom CSS classes to pass to the button */
  customclasses?: string | string[];
  /** Custom CSS classes to pass to the button label */
  customlabelclasses?: string | string[];
  /** The action that should be triggered when clicking the button */
  onClick(): unknown;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({ buttonShadow: { boxShadow: theme.shadows[1], border: 'transparant' } })
);

/**
 * Constructs a button using pre-defined Rijkswaterstaat styling
 * @remark Ref is forwarded to the Material-UI button to allow for wrapping this button with a Tooltip
 * @param props Props to pass to the button
 * @example
 * ```jsx
 * <Button label='button' onClick={() => undefined} variant='contained' color='primary' />
 * ```
 */

export const Button = memo(
  forwardRef<HTMLButtonElement, MUIButtonProps<'button', ButtonProps>>((props, ref) => {
    const classes = useStyles();

    return (
      <MUIButton
        {...props}
        ref={ref}
        data-qa={props['data-qa']}
        onClick={props.onClick}
        variant={props.variant}
        color={props.color}
        disabled={props.disabled}
        className={clsx(props.customclasses)}
        classes={{
          root: css.button,
          label: clsx(css.buttonLabel, props.customlabelclasses),
          disabled: css.buttonDisabled,
          containedPrimary: clsx(css.buttonPrimary, classes.buttonShadow),
          containedSecondary: clsx(css.buttonSecondary, classes.buttonShadow),
          outlined: clsx(css.buttonOutlined, classes.buttonShadow),
          outlinedPrimary: clsx(css.buttonOutlined, classes.buttonShadow),
          outlinedSecondary: clsx(css.buttonOutlinedSecondary, classes.buttonShadow)
        }}
        href={undefined}
      >
        {props.label}
      </MUIButton>
    );
  })
);
