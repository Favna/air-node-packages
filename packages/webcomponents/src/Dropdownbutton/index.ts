export { default as Dropdownbutton, DropdownbuttonProps } from './Dropdownbutton';