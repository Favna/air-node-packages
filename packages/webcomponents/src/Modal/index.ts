export { default as Modal, ModalProps, ModalQAs } from './Modal';