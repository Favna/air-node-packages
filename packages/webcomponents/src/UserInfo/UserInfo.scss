@import '../styles/vars';

.authInfo {
  position: absolute;
  right: 1rem;
  top: 2rem;
}

.relogin {
  text-transform: none;
}
