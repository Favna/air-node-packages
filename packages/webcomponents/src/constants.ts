import { ReactNode } from 'react';

export type label = string | ReactNode;
export type cellContent = label;
export type customCss = string | string[];
export type dataQa = string;

export const objectHasProperty = <O extends {}>(obj: O, prop: keyof O) => prop in obj && obj[prop];

/**
 * Split a string by its latest space character in a range from the character 0 to the selected one.
 * @param str The text to split.
 * @param length The length of the desired string.
 * @param char The character to split with
 */
export function splitText(str: string, length: number, char = ' ') {
  const x = str.substring(0, length).lastIndexOf(char);
  const pos = x === -1 ? length : x;

  return str.substring(0, pos);
}

/**
 * Split a text by its latest space character in a range from the character 0 to the selected one.
 * @param str The text to split.
 * @param length The length of the desired string.
 */
export function cutText(str: string, length: number) {
  if (str.length < length) return str;
  const cut = splitText(str, length - 3);
  if (cut.length < length - 3) return `${cut}...`;

  return `${cut.slice(0, length - 3)}...`;
}