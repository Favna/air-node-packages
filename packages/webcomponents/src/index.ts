export * from './ActionBar';
export * from './Button';
export * from './constants';
export * from './LoadingSkeleton';
export * from './Logo';
export * from './Modal';
export * from './ModalContent';
export * from './PaperButton';
export * from './SearchBar';
export * from './Table';
export * from './TableBodyCell';
export * from './TableHeaderCell';
export * from './TablePagination';
export * from './TablePaginationActions';
export * from './TableToolbar';
export * from './Tooltip';
export * from './ZeroWidthSpace';
export * from './Dropdownbutton';