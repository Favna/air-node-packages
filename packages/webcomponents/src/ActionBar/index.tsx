import { Typography, TypographyProps, Box } from '@material-ui/core';
import Grid from '@material-ui/core/Grid/Grid';
import classnames from 'classnames';
import React, { memo, ReactNode, useMemo } from 'react';
import { When } from 'react-if';
import Button from '../Button';
import css from './ActionBar.scss';

export interface ActionBarProps {
  /** The title to show in the ActionBar */
  title: ReactNode;
  /**
   * Whether an action button should be displayed in the ActionBar
   * @default false
   */
  shouldHaveButton?: boolean;
  /**
   * Whether the button should be disabled
   * Only relevant when `shouldHaveButton` is set to true
   * @default false
   */
  shouldDisableButton?: boolean;
  /** The text that should be shown in the button */
  buttonLabel?: string;
  /** Typography variant for the header to take */
  typographyVariant?: TypographyProps['variant'];
  /** Custom CSS classes to pass to the title */
  customclasses?: string | string[];
  /** Data-qa tag to apply action bar bounding box */
  'data-qa'?: string;
  /** The action that should be invoked when the button is clicked */
  buttonAction?: (...args: any[]) => any;
}

/**
 * Creates an action bar using pre-defined Rijkswaterstaat styling
 * @param props Props to pass to the actionbar
 * @example
 * ```jsx
 * <ActionBar title='Cool title' />
 * ```
 */
export const ActionBar = memo(
  ({
    title,
    'data-qa': dataQa,
    shouldDisableButton = false,
    shouldHaveButton = false,
    buttonAction,
    buttonLabel,
    typographyVariant = 'h1',
    customclasses
  }: ActionBarProps) => {
    const getTitle = useMemo<ActionBarProps['title']>(() => {
      if (typeof title === 'string') {
        return (
          <Typography
            variant={typographyVariant}
            data-qa='action-bar-title'
            color='textPrimary'
            className={classnames(css.actionHeader, customclasses)}
          >
            {title}
          </Typography>
        );
      }

      if (typeof title === 'function') return title();

      return title;
    }, [customclasses, title, typographyVariant]);

    return (
      <Box component='div' className={classnames(css.actionBar)} data-qa={dataQa}>
        <Grid container direction='row' justify='space-between' alignItems='center' className={css.actionGridLeft}>
          <Grid item key={1} xs={6}>
            {getTitle}
          </Grid>
          <When condition={Boolean(shouldHaveButton)}>
            <Grid item key={2} xs={5} sm={6} className={css.actionGridRight}>
              <Button
                data-qa='action-bar-button'
                variant='contained'
                color='primary'
                label={buttonLabel}
                onClick={() => (buttonAction ? buttonAction() : null)}
                disabled={Boolean(shouldDisableButton)}
              />
            </Grid>
          </When>
        </Grid>
      </Box>
    );
  }
);

export default ActionBar;
