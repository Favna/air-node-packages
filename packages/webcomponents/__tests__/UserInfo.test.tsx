import { toBeCalled } from '@rws-air/utils';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import { UserInfo } from '../src/UserInfo';

describe('UserInfo tests', () => {
  const mockRelogin = jest.fn();

  describe('Component Testing', () => {
    let userInfo: ShallowWrapper;

    beforeAll(() => {
      userInfo = shallow(<UserInfo email='john.connor@rws.nl' reloginText='relogin' onReloginClick={mockRelogin} />);
    });

    test('should call onReloginClick when clicking button', () => {
      const button = userInfo.find('[data-qa="re-login-link"]');

      button.simulate('click');

      toBeCalled(mockRelogin, 1);
    });
  });

  describe('Snapshot Testing', () => {
    test('Required Props', () => {
      const userInfo = shallow(
        <UserInfo email='john.connor@rws.nl' reloginText='relogin' onReloginClick={mockRelogin} />
      );
      expect(userInfo).toMatchSnapshot();
    });
  });
});
