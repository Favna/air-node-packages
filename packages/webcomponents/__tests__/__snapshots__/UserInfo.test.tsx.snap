// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInfo tests Snapshot Testing Required Props 1`] = `
<Styled(MuiBox)
  className="authInfo"
  component="div"
>
  <Styled(MuiBox)
    component="div"
    data-qa="logged-in-email"
  >
    <When
      condition={true}
    >
      <WithStyles(ForwardRef(Typography))
        align="center"
        color="textPrimary"
        variant="subtitle2"
      >
        john.connor@rws.nl
      </WithStyles(ForwardRef(Typography))>
    </When>
  </Styled(MuiBox)>
  <WithStyles(ForwardRef(Button))
    className="relogin"
    data-qa="re-login-link"
    onClick={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    size="small"
  >
    <WithStyles(ForwardRef(Typography))
      align="center"
      color="textPrimary"
      variant="subtitle2"
    >
      relogin
    </WithStyles(ForwardRef(Typography))>
  </WithStyles(ForwardRef(Button))>
</Styled(MuiBox)>
`;
